addons/web/static/src/views/form/form_controller.js

```diff
@@ -354,6 +366,9 @@ export class FormController extends Component {
         if (this.model.root.isVirtual) {
             this.env.config.historyBack();
         }
+        if (this.props.onDiscard) {
+            await this.props.onDiscard(this.model.root);
+        }
```

Comment:
Do you need to wait for the onDiscard promise? Are your callbacks async? If not, I would remove the await (here and in save()). Also, for this one, I would call it before the call to historyBack, as it basically destroys the view.
```js
await root.discard();
if (onDiscard) {
   (await) onDiscard();
}
if (isVirtual) {
   historyBack();
}

Reply:
Change done, thanks :)
For the `await`, it don't need it. Actually, I go in these callbacks through the `beforeExecuteAction` where I already wait nothing :man_shrugging: